### PR TITLE
A likely plot

### PR DESCRIPTION
**Pull request by rimangan**
Plotting the likelihood function for selection for an input VCF. Added as an option for the vcfAfs cmd. Also added testing for vcfAfs. Have a weird bug when running on large datasets, so a draft for now.

### CODE REVIEW

**Comment by rimangan** [comment]
After two months working on other parts of the project, today I am revisiting our likelihood plotting tools. I have a few plots from real data I can show at code review, but I can say beyond the tests here that this program is working with genome-size data. Ready for review

File: cmd/vcfAfs/vcfAfs.go [hunk 1]
```diff
@@ -11,27 +11,52 @@ import (
...
 	exception.FatalOnErr(err)
 	f := popgen.AfsToFrequency(*g)
 	out := fileio.EasyCreate(outFile)
-	defer out.Close()
 	for i := 0; i < len(f); i++ {
-		fmt.Fprintf(out, "%f\n", f[i])
+		_, err = fmt.Fprintf(out, "%f\n", f[i])
+		exception.PanicOnErr(err)
+	}
+	err = out.Close()
```
Comment (cshoben) [inline cmd/vcfAfs/vcfAfs.go:35]: I was thinking, this is going to just close the out file, or does it close both that file and one vcfFile (incoming file)?

File: cmd/vcfAfs/vcfAfs_test.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+	inputFile               string
+	expectedFile            string
+	unPolarized             bool
+	plotSelectionLikelihood string
+	leftBound               float64
+	rightBound              float64
+	numberOfPoints          int
+	integralError           float64
+}{
+	{"testdata/simulate.N100.S100.Seed19.Alpha0.01.vcf", "testdata/test.afs.txt", true, "", -10, 10, 100, 1e-5},
```
Comment (cshoben) [inline cmd/vcfAfs/vcfAfs_test.go:20]: Would it be good to have at least one other input instance to test? Maybe one is enough, idk?

File: popgen/plot.go [hunk 3]
```diff
@@ -8,139 +8,189 @@ import (
...
+}
+
+//PlotAfsLikelihood
+func PlotAfsLikelihood(afs Afs, outFile string, leftBound float64, rightBound float64, numPoints int, integralError float64, divergenceAscertainment bool, D int) {
+	var err error
+	var alpha float64
+	out := fileio.EasyCreate(outFile)
+
+	allN := findAllN(afs)
+	binomMap := BuildBinomCache(allN)
```
Comment (cshoben) [inline popgen/plot.go:110]: Where is this BuildBinomCache function coming from? Internal to golang?

**Review by cshoben — APPROVED** [review]
Had a question about the testing, but nothing to hold back an approval. Great job!

**Review by RavenYL — DISMISSED** [review]
I am wondering if the Settings struct is something we are using now as a standard across lots of file formats? I remember seeing it in bam or was it bed also

File: cmd/vcfAfs/vcfAfs.go (hunk 1 above)
Comment (rimangan) [reply cmd/vcfAfs/vcfAfs.go:35]: The opening and closing of the vcfFile is handled inside the popgen.VcfToAfs function call. This just closes the output file.

File: popgen/plot.go (hunk 3 above)
Comment (rimangan) [reply popgen/plot.go:110]: This is a part of the MCMC code. When we calculate the likelihood function for a particular allele frequency spectrum, we often have to calculate binomial probabilities from large numbers and use those values over and over again. This helper function pre-calculates all possible binomials we will need for a particular allele frequency spectrum and saves them to a cache as a runtime optimization

File: cmd/vcfAfs/vcfAfs.go [hunk 4]
```diff
@@ -11,27 +11,52 @@ import (
...
 }
 
 func main() {
+	var expectedNumArgs int = 2
 	var unPolarized *bool = flag.Bool("unPolarized", false, "vcfAfs creates polarized derived frequency spectra by default. When true, the cmd returns unpolarized site frequency spectra.")
+	var plotSelectionLikelihood *string = flag.String("plotSelectionLikelihood", "", "Plots the likelihood function for the selection parameter from this VCF to a specified output file.")
+	var leftBound *float64 = flag.Float64("leftBound", -10.0, "Left bound for plotting of the likelihood function.")
+	var rightBound *float64 = flag.Float64("rightBound", 10.0, "Right bound for plotting of the likelihood function.")
+	var numberOfPoints *int = flag.Int("numberOfPoints", 99, "Set the number of points at which to evaluate the likelihood function.")
+	var integralError *float64 = flag.Float64("integralError", 1e-5, "Sets the error threshold for integral calculations in the likelihood calculation.")
```
Comment (craiglowe) [inline cmd/vcfAfs/vcfAfs.go:57]: Should we decrease this based on your recent analysis?

File: popgen/plot.go [hunk 5]
```diff
@@ -8,139 +8,189 @@ import (
...
 	binomCache := BuildBinomCache(allN)
 	//DEBUG:fmt.Printf("Len BinomCache:%v. Entry at index n is of length: %v.\n", len(binomCache), len(binomCache[n]))
 
 	//a short header is written to the file
-	fmt.Fprintf(out, "Frequency\tProbability\n")
+	_, err = fmt.Fprintf(out, "Frequency\tProbability\n")
 	//for each possible allele frequency for a segregating site, we calculate the probability to the file
 	for i := 1; i < n; i++ {
-		fmt.Fprintf(out, "%v\t%e\n", i, AlleleFrequencyProbability(i, n, alpha, binomCache, integralError))
+		_, err = fmt.Fprintf(out, "%v\t%e\n", i, AlleleFrequencyProbability(i, n, alpha, binomCache, integralError))
```
Comment (craiglowe) [inline popgen/plot.go:96]: Might be good to check the error after this line

File: popgen/plot.go [hunk 6]
```diff
@@ -8,139 +8,189 @@ import (
...
+	allN := findAllN(afs)
+	binomMap := BuildBinomCache(allN)
+
+	_, err = fmt.Fprintf(out, "Alpha\tLikelihood\n")
+	exception.PanicOnErr(err)
+
+	for i := 0; i <= numPoints; i++ {
+		alpha = leftBound + (float64(i)/float64(numPoints))*(rightBound-leftBound)
+		if divergenceAscertainment {
+			_, err = fmt.Fprintf(out, "%e\t%e\n", alpha, AfsDivergenceAscertainmentFixedAlpha(afs, alpha, binomMap, D, integralError))
```
Comment (craiglowe) [inline popgen/plot.go:118]: might be good to check this error

File: popgen/plot.go [hunk 7]
```diff
@@ -8,139 +8,189 @@ import (
...
+		} else {
+			_, err = fmt.Fprintf(out, "%e\t%e\n", alpha, AfsLikelihoodFixedAlpha(afs, alpha, binomMap, integralError))
```
Comment (craiglowe) [inline popgen/plot.go:120]: might be good to check this error

**Review by craiglowe — COMMENTED** [review]
The plots coming out of this have been really informative!
Just a couple tiny comments

File: cmd/vcfAfs/vcfAfs.go [hunk 8]
```diff
@@ -11,27 +11,52 @@ import (
...
+	PlotSelectionLikelihood string
+	LeftBound               float64
+	RightBound              float64
+	NumberOfPoints          int
+	IntegralError           float64
+	DivergenceAscertainment bool
+	D                       int //size of the ascertainment subset
+}
+
+func vcfAfs(vcfFile string, outFile string, s Settings) {
```
[the remark this answers is not in the document]
Comment (rimangan) [reply cmd/vcfAfs/vcfAfs.go:25]: AFS appears a lot in the literature. I'm happy to change the name if we think that's appropriate though

**Review by craiglowe — APPROVED** [review]
Great work!

**Review by RavenYL — APPROVED** [review]

**Review by dasnellings — APPROVED** [review]

**Pull request merged**